library/std/src/sys/unix/ext/io.rs

```diff
@@ -45,6 +57,20 @@ pub trait FromRawFd {
     /// descriptor they are wrapping. Usage of this function could
     /// accidentally allow violating this contract which can cause memory
     /// unsafety in code that relies on it being true.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use std::fs::File;
+    /// use std::os::unix::io::{FromRawFd, IntoRawFd};
+    ///
+    /// fn main() -> std::io::Result<()> {
+    ///     let f = File::open("foo.txt")?;
+    ///     let raw_fd = f.into_raw_fd();
+    ///     let f = File::from_raw_fd(raw_fd);
```

Comment:
```suggestion
    ///     SAFETY: no other functions call `from_raw_fd`, so there is only one owner for the file descriptor.
    ///     let f = unsafe { File::from_raw_fd(raw_fd) };
```

Reply:
I think we can skip the "no other functions call `from_raw_fd`", I think the more important one is there is only one owner for the `raw_fd` which I believe implements `Copy`.

Reply:
That is not a sufficient condition. If you have multiple raw_fds and call `from_raw_fd` on both, that's undefined behavior even though `raw_fd` is Copy.

Reply:
I mean "there is only one owner for the file descriptor" is enough, it implies that the same `raw_fd` was never used more than once.

Reply:
That is _not_ a sufficient condition. It's fine to have this code:

```rust
let fd1 = f.as_raw_fd();
let fd2 = f.into_raw_fd();
let fd3 = fd1;
println!("{}", fd1); // note that fd1 can still be used
```
It is _not_ fine to have this code:
```rust
let fd1 = f.as_raw_fd();
let fd2 = fd1;
let f1 = File::from_raw_fd(fd1);
let f2 = File::from_raw_fd(fd2); // UNSOUND: double free
```

Reply:
"there is only one owner for the file descriptor" means the second example you gave is wrong since it have two owners now. Now `fd1` is owned by `f1` and `f2`. I wonder if we can just drop the `Copy` to prevent this from happening, but probably not people can still `Clone`.

Reply:
'has two owners' is a property of `from_raw_fd`, not of the type itself.

Reply:
I think we might have been talking past each other - did you mean to say 'no other _functions_' is the bit that should be changed? That was a typo, sorry - I meant to say 'nowhere else' or something like that.

I do think it's important to get across that `RawFd` _is_ copy and _can_ have multiple owners though, you just can't call `from_raw_fd` more than once.

Reply:
You can call `from_raw_fd` more than once but then you need to use `ManuallyDrop`. E.g. if you have a `RawFd` and want to do a `stat` syscall then instead of going through libc you can do this: `ManuallyDrop::new(File::from_raw_fd(fd)).metadata()`